Nimma usage generates very different results

**Describe the bug**
5.5.0 introduces an optional rule jsonpath optimizer: nimma. It seems to not properly parse some paths.

**To Reproduce**

**ruleset:**
```yaml
rules:
  read-scope-prefix-on-get:
    type: validation
    severity: warn
    recommended: true
    formats: ["oas3"]
    description: Scope should start with something
    message: Scope should start with something
    given: "$.paths.*[?(@property === 'get' || @property === 'options')].security.*.my_auth.*"
    then:
      function: pattern
      functionOptions:
        match: "^urn:"

  no-404-response:
    type: validation
    severity: info
    recommended: true
    formats: ["oas3"]
    description: Operation must not define a 404 response when no ressource identifier.
    message: The request must not define a 404 response.
    given: "$.paths[?(!@property.includes('_id}'))][?(@property === 'get' || @property === 'post' || @property === 'delete' || @property === 'options' || @property === 'patch' )].responses"
    then:
      field: "404"
      function: falsy
```

**document:**
```yaml
openapi: 3.0.0

paths:
  /any/resources/{resource}:
    get:
      responses:
        "200":
          description: All is good
        "404":
          description: All is nowhere"
  /resources:
    get: &dummy-endpoint
      responses:
        "200":
          description: All is good
      security:
        - my_auth:
            - urn:something
    patch:
      <<: *dummy-endpoint
      security:
        - my_auth:
            - Urn:wrong-but-should-go-unnoticed
```
**without nimma:**
```bash
$ node -r ts-node/register -r tsconfig-paths/register src/cli/index.ts lint -r ./repro/nimma.ruleset.yaml ./repro/nimma.openapi.yaml
Missing baseUrl in compilerOptions. tsconfig-paths will be skipped
OpenAPI 3.x detected

c:/spectral/repro/nimma.openapi.yaml
 9:15  information  no-404-response  The request must not define a 404 response.

✖ 1 problem (0 errors, 0 warnings, 1 info, 0 hints)

```

**activating nimma:**
```bash
$ node -r ts-node/register -r tsconfig-paths/register src/cli/index.ts lint -r ./repro/nimma.ruleset.yaml ./repro/nimma.openapi.yaml
Missing baseUrl in compilerOptions. tsconfig-paths will be skipped
OpenAPI 3.x detected

c:/_work/spectral/repro/nimma.openapi.yaml
 23:15  warning  read-scope-prefix-on-get  Scope should start with something

✖ 1 problem (0 errors, 1 warning, 0 infos, 0 hints)
```

**Expected behavior**
Using nimma or not should not change the return results.

**Environment (remove any that are not applicable):**
 - Library version: develop branch
 - OS: Win10
 - Browser: N/A

--- a/jest.config.js
+++ b/jest.config.js
@@ -8,7 +8,9 @@ const projectDefault = {
   preset: 'ts-jest',
   moduleNameMapper: {
     ...mapValues(pathsToModuleNameMapper(compilerOptions.paths), v => path.join(__dirname, v)),
-    '@stoplight/spectral-test-utils': '<rootDir>/test-utils/node/index.ts',
+    '^@stoplight/spectral-test-utils$': '<rootDir>/test-utils/node/index.ts',
+    '^nimma/fallbacks$': '<rootDir>/node_modules/nimma/dist/cjs/fallbacks/index.js',
+    '^nimma/legacy$': '<rootDir>/node_modules/nimma/dist/legacy/cjs/index.js',
   },
   testEnvironment: 'node',
   globals: {

--- a/karma.conf.ts
+++ b/karma.conf.ts
@@ -37,6 +37,9 @@ module.exports = (config: Config): void => {
         resolve: {
           alias: {
             '@stoplight/spectral-test-utils': require.resolve('./test-utils/browser/index.js'),
+            nimma: require.resolve('./node_modules/nimma/dist/legacy/cjs/index.js'),
+            'nimma/fallbacks': require.resolve('./node_modules/nimma/dist/legacy/cjs/fallbacks/index.js'),
+            'nimma/legacy': require.resolve('./node_modules/nimma/dist/legacy/cjs/index.js'),
             'node-fetch': require.resolve('./__karma__/fetch'),
             fs: require.resolve('./__karma__/fs'),
             process: require.resolve('./__karma__/process'),

--- a/packages/core/package.json
+++ b/packages/core/package.json
@@ -36,10 +36,11 @@
     "blueimp-md5": "2.18.0",
     "expression-eval": "4.0.0",
     "json-schema": "0.3.0",
-    "jsonpath-plus": "5.0.7",
+    "jsonpath-plus": "6.0.1",
     "lodash": "~4.17.21",
+    "lodash.topath": "^4.5.2",
     "minimatch": "3.0.4",
-    "nimma": "0.0.0",
+    "nimma": "0.1.1",
     "tslib": "~2.3.0"
   },
   "devDependencies": {

--- a/packages/core/src/ruleset/rule/rule.ts
+++ b/packages/core/src/ruleset/rule/rule.ts
@@ -1,11 +1,9 @@
 import { JsonPath, Optional } from '@stoplight/types';
-import { JSONPathExpression } from 'nimma';
 import { dirname, relative } from '@stoplight/path';
 import { DiagnosticSeverity } from '@stoplight/types';
 import { pathToPointer } from '@stoplight/json';
 
 import { getDiagnosticSeverity, DEFAULT_SEVERITY_LEVEL } from '../utils/severity';
-import { IGivenNode } from '../../types';
 import { Ruleset } from '../ruleset';
 import { Format } from '../format';
 import { HumanReadableDiagnosticSeverity, IRuleThen, RuleDefinition } from '../types';
@@ -45,12 +43,6 @@ export class Rule implements IRule {
   public documentationUrl: string | null;
   #then!: IRuleThen[];
   #given!: string[];
-
-  public expressions?: JSONPathExpression[] | null;
-
-  public get isOptimized(): boolean {
-    return Array.isArray(this.expressions);
-  }
 
   constructor(
     public readonly name: string,
@@ -193,38 +185,8 @@ export class Rule implements IRule {
     return false;
   }
 
-  public optimize(): boolean {
-    if (this.expressions !== void 0) return this.isOptimized;
-
-    try {
-      this.expressions = this.given.map(given => {
-        const expr = new JSONPathExpression(given, stub, stub);
-        if (expr.matches === null) {
-          throw new Error(`Rule "${this.name}": cannot optimize ${given}`);
-        }
-
-        return expr;
-      });
-    } catch {
-      this.expressions = null;
-    }
-
-    return this.isOptimized;
-  }
-
   public clone(): Rule {
     return new Rule(this.name, this.definition, this.owner);
-  }
-
-  public hookup(cb: (rule: Rule, node: IGivenNode) => void): void {
-    for (const expr of this.expressions!) {
-      expr.onMatch = (value, path): void => {
-        cb(this, {
-          path,
-          value,
-        });
-      };
-    }
   }
 
   public toJSON(): StringifiedRule {
@@ -247,8 +209,4 @@ export class Rule implements IRule {
       owner: this.owner.id,
     };
   }
-}
-
-function stub(): void {
-  // nada
 }

--- a/packages/core/src/runner/runner.ts
+++ b/packages/core/src/runner/runner.ts
@@ -1,53 +1,15 @@
-import { JSONPath, JSONPathCallback } from 'jsonpath-plus';
-import { isObject } from 'lodash';
-import { JSONPathExpression, traverse } from 'nimma';
-
 import { IDocument } from '../document';
 import { DocumentInventory } from '../documentInventory';
-import { IGivenNode, IRuleResult } from '../types';
+import { IRuleResult } from '../types';
 import { ComputeFingerprintFunc, prepareResults } from '../utils';
 import { lintNode } from './lintNode';
 import { RunnerRuntime } from './runtime';
 import { IRunnerInternalContext } from './types';
-import { Rule } from '../ruleset/rule/rule';
 import { Ruleset } from '../ruleset/ruleset';
-import { toPath } from 'lodash';
-
-const runRule = (context: IRunnerInternalContext, rule: Rule): void => {
-  const target = rule.resolved ? context.documentInventory.resolved : context.documentInventory.unresolved;
-
-  for (const given of rule.given) {
-    // don't have to spend time running jsonpath if given is $ - can just use the root object
-    if (given === '$') {
-      lintNode(
-        context,
-        {
-          path: ['$'],
-          value: target,
-        },
-        rule,
-      );
-    } else if (isObject(target)) {
-      JSONPath({
-        path: given,
-        json: target,
-        resultType: 'all',
-        callback: (result => {
-          lintNode(
-            context,
-            {
-              // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access,@typescript-eslint/no-unsafe-assignment,@typescript-eslint/no-unsafe-call
-              path: toPath(result.path.slice(1)),
-              // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
-              value: result.value,
-            },
-            rule,
-          );
-        }) as JSONPathCallback,
-      });
-    }
-  }
-};
+import Nimma, { Callback } from 'nimma/legacy'; // legacy = Node v12, nimma without /legacy supports only 14+
+import { jsonPathPlus } from 'nimma/fallbacks';
+import { isPlainObject } from '@stoplight/json';
+import { isError } from 'lodash';
 
 export class Runner {
   public readonly results: IRuleResult[];
@@ -77,43 +39,35 @@ export class Runner {
       promises: [],
     };
 
-    const relevantRules = Object.values(rules).filter(
-      rule => rule.enabled && rule.matchesFormat(documentInventory.formats),
-    );
-
-    const optimizedRules: Rule[] = [];
-    const optimizedUnresolvedRules: Rule[] = [];
-    const unoptimizedRules: Rule[] = [];
-
-    const traverseCb = (rule: Rule, node: IGivenNode): void => {
-      lintNode(runnerContext, node, rule);
+    const enabledRules = Object.values(rules).filter(rule => rule.enabled);
+    const relevantRules = enabledRules.filter(rule => rule.matchesFormat(documentInventory.formats));
+    const callbacks: { resolved: Record<string, Callback[]>; unresolved: Record<string, Callback[]> } = {
+      resolved: {},
+      unresolved: {},
     };
 
     for (const rule of relevantRules) {
-      if (!rule.isOptimized) {
-        unoptimizedRules.push(rule);
-        continue;
+      for (const given of rule.given) {
+        const cb: Callback = (scope): void => {
+          lintNode(runnerContext, scope, rule);
+        };
+
+        (callbacks[rule.resolved ? 'resolved' : 'unresolved'][given] ??= []).push(cb);
       }
-
-      if (rule.resolved) {
-        optimizedRules.push(rule);
-      } else {
-        optimizedUnresolvedRules.push(rule);
-      }
-
-      rule.hookup(traverseCb);
     }
 
-    if (optimizedRules.length > 0) {
-      traverse(Object(runnerContext.documentInventory.resolved), optimizedRules.flatMap(pickExpressions));
-    }
+    execute(
+      runnerContext.documentInventory.resolved,
+      callbacks.resolved,
+      relevantRules.flatMap(r => (r.resolved ? r.given : [])),
+    );
 
-    if (optimizedUnresolvedRules.length > 0) {
-      traverse(Object(runnerContext.documentInventory.unresolved), optimizedUnresolvedRules.flatMap(pickExpressions));
-    }
-
-    for (const rule of unoptimizedRules) {
-      runRule(runnerContext, rule);
+    if (Object.keys(callbacks.unresolved).length > 0) {
+      execute(
+        runnerContext.documentInventory.unresolved,
+        callbacks.unresolved,
+        relevantRules.flatMap(r => (!r.resolved ? r.given : [])),
+      );
     }
 
     this.runtime.emit('beforeTeardown');
@@ -132,6 +86,46 @@ export class Runner {
   }
 }
 
-function pickExpressions({ expressions }: Rule): JSONPathExpression[] {
-  return expressions!;
+function execute(input: unknown, callbacks: Record<string, Callback[]>, jsonPathExpressions: string[]): void {
+  if (!isPlainObject(input) && !Array.isArray(input)) {
+    for (const cb of callbacks.$ ?? []) {
+      cb({
+        path: [],
+        value: input,
+      });
+    }
+
+    return;
+  }
+
+  try {
+    const nimma = new Nimma(jsonPathExpressions, {
+      fallback: jsonPathPlus,
+      unsafe: false,
+      output: 'auto',
+    });
+
+    nimma.query(
+      input,
+      Object.entries(callbacks).reduce<Record<string, Callback>>((mapped, [key, cbs]) => {
+        mapped[key] = scope => {
+          for (const cb of cbs) {
+            cb(scope);
+          }
+        };
+
+        return mapped;
+      }, {}),
+    );
+  } catch (e) {
+    if (isAggregateError(e) && e.errors.length === 1) {
+      throw e.errors[0];
+    } else {
+      throw e;
+    }
+  }
+}
+
+function isAggregateError(maybeAggregateError: unknown): maybeAggregateError is Error & { errors: unknown[] } {
+  return isError(maybeAggregateError) && maybeAggregateError.constructor.name === 'AggregateError';
 }

--- a/packages/core/src/types/spectral.ts
+++ b/packages/core/src/types/spectral.ts
@@ -4,7 +4,6 @@ import type { Resolver } from '@stoplight/spectral-ref-resolver';
 
 export interface IConstructorOpts {
   resolver?: Resolver;
-  useNimma?: boolean;
 }
 
 export interface IRunOpts {

--- a/tsconfig.json
+++ b/tsconfig.json
@@ -10,7 +10,10 @@
       "@stoplight/spectral-ref-resolver": ["packages/ref-resolver/src/index.ts"],
       "@stoplight/spectral-runtime": ["packages/runtime/src/index.ts"],
       "@stoplight/spectral-ruleset-migrator": ["packages/ruleset-migrator/src/index.ts"],
-      "@stoplight/spectral-rulesets": ["packages/rulesets/src/index.ts"]
+      "@stoplight/spectral-rulesets": ["packages/rulesets/src/index.ts"],
+
+      "nimma/fallbacks": ["node_modules/nimma/dist/cjs/fallbacks/"],
+      "nimma/legacy": ["node_modules/nimma/dist/legacy/cjs/"]
     },
     "moduleResolution": "node",
     "target": "ES2019",

